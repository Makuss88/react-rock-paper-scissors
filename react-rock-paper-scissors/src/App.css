.wrapper {
  display: flex;  
  flex-flow: row wrap;
  text-align: center;
  justify-content: center;
  align-items: center;
}

.wrapper > * {
  padding: 24px;
  flex: 1 100%;
}

.movesWrapper {
  display: flex;
  align-items: center;
  justify-content: space-evenly;
}

.aside {
  width: 100px;
}

@media all and (min-width: 600px) {
  .aside { flex: 1 auto; }
}

@media all and (min-width: 800px) {
  .main    { flex: 3 0px; }
  .aside-1 { order: 1; } 
  .main    { order: 2; }
  .aside-2 { order: 3; }

  .wrapper {
    margin: 106px 0 0; 
    margin: 18vh 0 0; 
  }
}

.footer {
  position: absolute;
  bottom: 0;
  width: 100%;
  background: #ff6347;
  padding: 24px 0;
}

.button {
  background-color: #ffffff;
  margin: 0 12px;
  border: none;
  height: 60px;
  outline: none;
  width: 60px;
  border-radius: 50%;
  box-shadow: 0 0 0 #ffffff;
  animation: pulse2s 2s infinite;
}

.icon {
  font-size: 40px;
  cursor: pointer;
  padding-top: 6px;
  display: inline-block;
  vertical-align: middle;
}

@keyframes pulse2s {
  0% {
    -moz-box-shadow: 0 0 0 0 rgba(255,255,255, 0.4);
    box-shadow: 0 0 0 0 rgba(255,255,255, 0.4);
  }
  70% {
      -moz-box-shadow: 0 0 0 8px rgba(255,255,255, 0);
      box-shadow: 0 0 0 8px rgba(255,255,255, 0);
  }
  100% {
      -moz-box-shadow: 0 0 0 0 rgba(255,255,255, 0);
      box-shadow: 0 0 0 0 rgba(255,255,255, 0);
  }
}