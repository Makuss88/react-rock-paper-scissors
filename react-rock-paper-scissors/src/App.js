import React, { Component } from 'react';
import Move from './components/Move';
import ResultLabel from './components/ResultLabel';
import './App.css';

export const winnerMoves = {
	'Rock': {
    wins: ['Scissors'],
    icon: '👊',
	},
  'Paper': {
    wins: ['Rock'],
    icon: '✋️',
	},
	'Scissors': {
    wins: ['Paper'],
    icon: '✌️',
	},
};

const possibleMoves = Object.keys(winnerMoves);

export const getRandomNumber = () => {
  return Math.floor(Math.random() * possibleMoves.length + 1);
};

class App extends Component {
  constructor(props) {
    super(props);
    this.state = {
        userMove: null,
        computerMove: null,
        winner: null,
        userWins: 0,
        computerWins: 0,
    };
  }

  getRandomMove = () => {
    const randomNumber = getRandomNumber();
    let computerRandomMove = {
      text: '',
      icon: '',
    };

    switch(randomNumber) {
        case 1:
          computerRandomMove = {
            text: 'Rock',
            icon: winnerMoves['Rock'].icon,
          }
          break;
        case 2:
          computerRandomMove = {
            text: 'Paper',
            icon: winnerMoves['Paper'].icon,
          }
          break;
        default:
          computerRandomMove = {
            text: 'Scissors',
            icon: winnerMoves['Scissors'].icon,
          }
          break;
    }
    return computerRandomMove;
  };

  getWinner = (userMove, computerMove) => {
    if (userMove === computerMove) return 0;
    return winnerMoves[userMove].wins.some(wins => wins === computerMove) ? 1 : 2;
  }

  getUserMoveAndUpdateState = (userMoveText) => {
    const nextComputerMove = this.getRandomMove();
    const nextUserMove = {
      text: userMoveText,
      icon: winnerMoves[`${userMoveText}`].icon,
    }
    const nextWinner = this.getWinner(nextUserMove.text, nextComputerMove.text);

    this.setState((prevState) => {
      return {
        userMove: nextUserMove,
        computerMove: nextComputerMove,
        winner: nextWinner,
        userWins: (nextWinner === 1) ? prevState.userWins + 1 : prevState.userWins,
        computerWins: (nextWinner === 2) ? prevState.computerWins + 1 : prevState.computerWins,
      }
    });
  }

  render() {
    return (
        <div className="wrapper">
          <aside className="aside aside-1">
            <h2>You</h2>
            <p>{this.state.userWins} point(s)</p>
          </aside>
          <article className="main">
            <div className="movesWrapper">
              <Move selectedOption = { this.state.userMove || '' } />
                <span>vs.</span>
              <Move selectedOption = { this.state.computerMove || '' } />
            </div>
            <ResultLabel winner={ this.state.winner } />
          </article>
          <aside className="aside aside-2">
            <h2>Computer</h2>
            <p> {this.state.computerWins} point(s)</p>
          </aside>
          <footer className="footer">
          { 
              possibleMoves.map((move) => {
                return (
                  <button className="button" key={move} onClick={ () => this.getUserMoveAndUpdateState(move) }>
                    <span className="icon" role="img" aria-label={ move }> { winnerMoves[move].icon }</span>
                  </button>);
              })
            }
          </footer>
        </div>
    );
  }
}

export default App;
