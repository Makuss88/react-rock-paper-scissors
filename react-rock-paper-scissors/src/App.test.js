import React from 'react';
import ReactDOM from 'react-dom';
import Enzyme, { shallow, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import toJson from 'enzyme-to-json';
import { mockRandom, resetMockRandom } from 'jest-mock-random';
import App, { getRandomNumber, getWinner } from './App';

Enzyme.configure({ adapter: new Adapter() });

describe('App', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<App />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('renders initial state before game start', () => {
    const wrapper = shallow(<App />); 
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});

describe('App after a user play', () => {
  it('displays the user move', () => {
    const mockUserMove = {
        text: 'Rock',
        icon: 'ðŸ‘Š',
    }
    const wrapper = mount(<App />); 
    const selectedIcon = wrapper.find(`.${mockUserMove.text}`);

    selectedIcon.simulate('click');
    expect(wrapper.state().userMove).toEqual(mockUserMove);
    expect(wrapper.find('.movedIcon').first().text()).toEqual(mockUserMove.icon);
  });

  it('displays the computer move', () => {
    mockRandom([0]);

    const mockUserMove = {
        text: 'Rock',
        icon: 'ðŸ‘Š',
    }
    const mockComputerMove = {
      text: 'Rock',
      icon: 'ðŸ‘Š',
    }
    const wrapper = mount(<App />); 
    const selectedIcon = wrapper.find(`.${mockUserMove.text}`);

    selectedIcon.simulate('click');
    expect(wrapper.state().computerMove).toEqual(mockComputerMove);
    expect(wrapper.find('.movedIcon').last().text()).toEqual(mockComputerMove.icon);

    resetMockRandom();
  });

  it('updates user points when user wins', () => {
    mockRandom([0]);

    const mockUserMove = {
        text: 'Paper',
    }
    const wrapper = shallow(<App />); 
    const selectedIcon = wrapper.find(`.${mockUserMove.text}`);

    selectedIcon.simulate('click');
    expect(wrapper.state().winner).toEqual(1);
    expect(wrapper.state().userWins).toEqual(1);
    expect(wrapper.state().computerWins).toEqual(0);
  
    resetMockRandom();
  });

  it('doesn\'t update points if tied', () => {
    mockRandom([0.34]);

    const mockUserMove = {
        text: 'Paper',
    }
    const wrapper = shallow(<App />); 
    const selectedIcon = wrapper.find(`.${mockUserMove.text}`);
  
    selectedIcon.simulate('click');
    expect(wrapper.state().winner).toEqual(0);
    expect(wrapper.state().userWins).toEqual(0);
    expect(wrapper.state().computerWins).toEqual(0);
  
    resetMockRandom();
  });

  it('updates computer points when computer wins', () => {
    mockRandom([0.67]);

    const mockUserMove = {
        text: 'Paper',
    }
    const wrapper = shallow(<App />); 
    const selectedIcon = wrapper.find(`.${mockUserMove.text}`);

    selectedIcon.simulate('click');

    expect(wrapper.state().winner).toEqual(2);
    expect(wrapper.state().userWins).toEqual(0);
    expect(wrapper.state().computerWins).toEqual(1);

    resetMockRandom();
  });
});

describe('getRandomNumber', () => {
  it('returns a random number between x and y', () => {
    const x = 1;
    const y = 3;
  
    expect(getRandomNumber(y,x)).toBeGreaterThanOrEqual(x);
    expect(getRandomNumber(y,x)).toBeLessThanOrEqual(y);
  })
});

describe('getWinner', () => {
  it('computes tied result for equal moves', () => {
    const mockUserMove = {
        text: 'Rock',
    }
    const mockComputerMove = {
      text: 'Rock',
    }
    const winner = getWinner(mockUserMove.text, mockComputerMove.text);

    expect(winner).toEqual(0);
  });

  it('computes user as winner', () => {
    const mockUserMove = {
        text: 'Paper',
    }
    const mockComputerMove = {
      text: 'Rock',
    }
    const winner = getWinner(mockUserMove.text, mockComputerMove.text);

    expect(winner).toEqual(1);
  });

  it('computes computer as winner', () => {
    const mockUserMove = {
        text: 'Paper',
    }
    const mockComputerMove = {
      text: 'Scissors',
    }
    const winner = getWinner(mockUserMove.text, mockComputerMove.text);

    expect(winner).toEqual(2);
  });
});