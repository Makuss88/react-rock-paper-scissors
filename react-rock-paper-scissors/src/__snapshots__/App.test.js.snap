// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders initial state before game start 1`] = `
<div
  className="wrapper"
>
  <aside
    className="aside aside-1"
  >
    <h2>
      You
    </h2>
    <p>
      0
       point(s)
    </p>
  </aside>
  <article
    className="main"
  >
    <div
      className="movesWrapper"
    >
      <Move
        selectedOption=""
      />
      <span>
        vs.
      </span>
      <Move
        selectedOption=""
      />
    </div>
    <ResultLabel
      winner={null}
    />
  </article>
  <aside
    className="aside aside-2"
  >
    <h2>
      Computer
    </h2>
    <p>
       
      0
       point(s)
    </p>
  </aside>
  <footer
    className="footer"
  >
    <button
      className="button Rock"
      key="Rock"
      onClick={[Function]}
    >
      <span
        aria-label="Rock"
        className="icon"
        role="img"
      >
         
        👊
      </span>
    </button>
    <button
      className="button Paper"
      key="Paper"
      onClick={[Function]}
    >
      <span
        aria-label="Paper"
        className="icon"
        role="img"
      >
         
        ✋️
      </span>
    </button>
    <button
      className="button Scissors"
      key="Scissors"
      onClick={[Function]}
    >
      <span
        aria-label="Scissors"
        className="icon"
        role="img"
      >
         
        ✌️
      </span>
    </button>
  </footer>
</div>
`;
